### ISSUE TITLE
PAR-1 — Continuous in-process legacy-vs-v1 parity harness

### ISSUE THREAD

**Issue by aacostadiaz** [ACEsuit/mace#1572]
**Depends on:** INF-2 (#1551), INF-3 (#1552), FM-00 (#1565) · **Blocks:** PAR-2 (#1573), RET-1 (#1597)

**Prerequisite:** the Phase 0 characterization suite is committed — goldens, fixtures and the single-source tolerance table under `tests/golden/`, characterization tests under `tests/unit/`. References of the form `P0-N` name that work.

**Context:** The reforge builds the v1 stack (`packages/{mace-core,mace-torch}`) alongside a byte-frozen legacy `mace/` package, and the two stacks never import each other — the one-way import direction is machine-enforced (INF-3 (#1552)). This ticket builds `tests/parity/`, the **second and last module on the dual-import allowlist** (the first is `mace_launcher`, INF-2 (#1551)): the only test code allowed to load both stacks and compare them numerically. It is what turns frozen legacy into a **live, in-process differential oracle** — every migrated capability is diffed against the real legacy implementation on every PR, not only against committed JSON goldens. Once PAR-1 (#1572) is green for a capability, that capability's `--engine` default can flip from `legacy` to `v1` with opt-out; the retirement PRs (RET-1 (#1597)…) later degrade each in-process parity test to its frozen golden when the legacy counterpart is deleted.

Two failure modes motivate the two hard parts of this ticket:

- **Global process-state contamination.** Importing legacy mutates the process: `mace/__init__.py` sets the env var `TORCH_FORCE_NO_WEIGHTS_ONLY_LOAD=1`; e3nn's `set_optimization_defaults` is process-wide; torch's default dtype and the torch/numpy RNG state are global. Run legacy then v1 naively in one process and the comparison is silently invalid.
- **The training-numerics gap.** A comparison against a JSON of final errors cannot see a gradient bug the size of init noise. Closing it needs a one-step, init-agnostic gradient parity test — and P0-6 has already committed the legacy side as a golden.

**Interface & constraints:**

1. **Shared machinery.** The harness consumes P0-1's framework-agnostic `tests/golden/harness.py` (numpy/ase/json only; it never imports `mace`, which is why code living outside the legacy zone can use it), the fixtures in `tests/golden/fixtures/` (triclinic periodic bulk ≥3 species, water cluster, isolated atom, ~0.6 Å dimer, mixed-PBC), and the two committed tiny anchors — a `ScaleShiftMACE` and a plain `MACE`, both trained with ZBL pair repulsion, because the pair-repulsion term is applied *scaled inside* `ScaleShiftMACE` (`mace/modules/models.py:545`) but *added outside* the scale in plain `MACE` (`mace/modules/models.py:361-362`); only exercising both turns that divergence into a numeric gate. The mixed-PBC fixture earns its place twice over now: it is also the case that separates DATA-3 (#1571)'s three effective-cell regimes.
2. **Tolerance table** — single source of truth in `tests/golden/harness.py`, imported, never copied: fp64 CPU cross-machine `1e-6 eV` / `1e-6 eV/Å`; cueq/oeq fp64 `1e-5`; fp32 `1e-3` relative. The table is **edit-locked**: a tolerance change is its own standalone, reviewed PR — never part of this or any feature ticket. Goldens are likewise never regenerated here.
3. **Process-state isolation (hard requirement).** Between the legacy run and the v1 run the harness must either (a) **snapshot and restore all global state**: relevant env vars (`TORCH_FORCE_NO_WEIGHTS_ONLY_LOAD`), `torch.get_default_dtype()`, `e3nn.get/set_optimization_defaults` (legacy-side), and torch + numpy RNG state; or (b) **fork a subprocess** per stack and diff serialized outputs. Either mode is acceptable; the isolation itself must be asserted by a test, not assumed. Note develop has already had to defend the dtype half of this on its own account — `tests/extensions/magnetic/test_magmace.py:983,1002` assert the magnetic calculator neither changes the global dtype nor breaks under a foreign one — so the snapshot list is a superset of a known-live hazard, not a hypothetical.
4. **Output parity.** The same tiny fixtures run through frozen legacy and through v1 (anchor checkpoints enter v1 via the FM-00 (#1565) in-process converter, whose neutral-format conversion is already golden-gated), asserting **fp64 energy / forces / stress / dipole parity** against the tolerance table. One conformance item pinned here: the v1 reference `segment_reduce` must implement the **same sequential-per-segment accumulation order as the legacy vendored `scatter_sum` on CPU** (`mace/tools/scatter.py`, excluded from lint/coverage precisely because it is vendored) — otherwise atom→graph reductions differ below tolerance-visibility in ways that surface later at fp32.
5. **Black-box re-run.** The **68 black-box test functions** (54 in `tests/workflows`, 14 in `tests/integrations`) re-run under `--engine v1` for migrated capabilities and **skip cleanly** otherwise (the launcher's clear "capability X not yet available on v1 engine" error from INF-2 (#1551) is the skip trigger, not an xfail soup). The enabling edit is made **once** in `tests/helpers.py`: it currently hardcodes `run_train = REPO_ROOT / "mace" / "cli" / "run_train.py"` (`tests/helpers.py:31`, with `preprocess_data` beside it at `:32`) and builds `cmd = [sys.executable, str(script)]` (`:201`); route it through the `mace_launcher` entry point instead, engine selected by env (`MACE_ENGINE`). The tests themselves are not rewritten. Respect the integration tier split while doing it: `tests/integrations` is two tiers (contract without a binary, and the `bin_lammps` real tier — see `tests/integrations/README.md`), and the real tier runs only in nightly's `lammps-real` job, so the v1 re-run inherits the same marker gating rather than forcing a binary into per-PR CI.
6. **Single-step, init-agnostic gradient parity (per-PR).** Two stages, in order:
   - **First reproduce the P0-6 legacy golden**: with the fixed seed and the committed anchor weights, one forward+backward of an energy+forces loss on the tiny fixture must reproduce the committed `d(loss)/d(params)` references (`tests/golden/references/tiny_{scaleshift,mace}_train_step_grad_fp64.json`) on the frozen legacy stack — so a v1 gradient regression fails against a **committed number**, not only against the other live stack (which could drift in lockstep).
   - **Then diff legacy vs v1**: load **byte-identical weights** into both stacks (possible because both run the anchor's own basis/layout; the FM-00 (#1565) mapping is exact for the anchors), run one forward+backward of the same energy+forces loss — this exercises the `create_graph=True` second-derivative path that force training backpropagates through — and assert `d(loss)/d(params)` allclose at fp64.
   - Plus a **model-level force+stress `gradgradcheck`** on the tiny fixture (v1 side).
7. **Cadence.** The harness runs **per-PR** (it is the merge gate for every migrated capability), wired into `ci-core.yaml` alongside `lint` / `unit` / `workflows` / `coverage`, so it must stay tiny-fixture-fast on CPU. Property/fuzz inputs and the fp32 bit-reproduction golden are nightly work (PAR-2 (#1573)), not here.

**Task:**
1. Scaffold `tests/parity/` (conftest + runner): the process-state snapshot/restore fixture or the forked-subprocess runner (constraint 3), loading legacy and v1 in a controlled order; register `tests/parity/` as the second allowlisted module in the INF-3 (#1552) import-linter config + audit-hook (and nothing else).
2. Output-parity tests: both P0-1 anchors through both stacks on all tiny fixtures; E/F/stress (dipole where the model provides it) compared via `tests/golden/harness.py` at the fp64 row. Include the `segment_reduce` CPU accumulation-order conformance check.
3. Black-box re-run: edit `tests/helpers.py` once to invoke the launcher (engine by `MACE_ENGINE`); make the 68 functions runnable under `--engine v1`, skipping cleanly on the launcher's not-yet-migrated error and preserving the `bin_lammps` tier gating.
4. Gradient parity per constraint 6: legacy reproduction of the P0-6 golden, then legacy-vs-v1 `d(loss)/d(params)` allclose at fp64 with byte-identical weights, then the model-level force+stress `gradgradcheck`.
5. Wire `tests/parity` into the per-PR CI job set (`ci-core.yaml`), so a parity break fails PRs from the day the harness lands.

**Out of scope:** nightly property/fuzz parity, the frozen fp32 bit-reproduction golden, and per-input-class coverage counters (PAR-2 (#1573)); the legacy gradient-golden characterization itself (P0-6, already committed); flipping any `--engine` default to `v1` (a per-capability decision gated on this harness being green); GPU-backend parity (BKD-3 (#1567)); any tolerance or golden edit.

**Acceptance criteria:**
- [ ] fp64 E/F/stress (and dipole where available) parity green for both anchors on all tiny fixtures at the edit-locked tolerance row, legacy and v1 loaded in the same process (or forked subprocesses).
- [ ] A test asserts process-state restoration: env vars, default dtype, e3nn optimization defaults, and torch/numpy RNG state are identical before and after a legacy run inside the harness (or the subprocess mode is exercised).
- [ ] The P0-6 `d(loss)/d(params)` goldens are reproduced on the frozen legacy stack from inside the harness for both anchors.
- [ ] Legacy-vs-v1 `d(loss)/d(params)` allclose at fp64 with byte-identical weights on an energy+forces loss (`create_graph=True` path exercised); model-level force+stress `gradgradcheck` green on the tiny fixture.
- [ ] The 68 black-box test functions run under `--engine v1`: migrated capabilities pass, unmigrated ones skip cleanly (no failures, no silent passes); `tests/helpers.py` routes through the launcher in one place; the `bin_lammps` tier stays nightly-only.
- [ ] The v1 reference `segment_reduce` matches legacy CPU sequential-per-segment accumulation order.
- [ ] `tests/parity/` is on the import-direction allowlist; no other module gained dual-import rights; no golden or tolerance file changed in this PR.
- [ ] The parity job runs per-PR in CI.

**Verify:**
```bash
python -m pytest tests/parity -v                                   # output + gradient parity, state isolation
python -m pytest tests/parity -k "gradient" -v                     # P0-6 golden reproduction + legacy-vs-v1 diff
MACE_ENGINE=v1 python -m pytest tests/workflows tests/integrations -n 2   # migrated pass, rest skip cleanly
lint-imports                                                       # allowlist: mace_launcher + tests/parity only
```
**Review focus:** completeness of the process-state snapshot (one leaked global silently invalidates the oracle — enumerate the state, don't sample it), and the gradient-parity construction: byte-identical weights on both sides, the loss actually traversing `create_graph=True`, and the legacy leg failing against the committed golden rather than only against live v1.

---
